internal/transport/grpchttp2/http2bridge_test.go

```diff
@@ -0,0 +1,468 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpchttp2
+
+import (
+	"errors"
+	"fmt"
+	"testing"
+
+	"golang.org/x/net/http2/hpack"
+)
+
+// testConn is a test utility which provides an io.Writer and io.Reader
+// interface and access to its internal buffers for testing.
+type testConn struct {
+	wbuf []byte
+	rbuf []byte
+}
+
+func (c *testConn) Write(p []byte) (int, error) {
+	c.wbuf = append(c.wbuf, p...)
+	return len(p), nil
+}
+
+func (c *testConn) Read(p []byte) (int, error) {
+	n := copy(p, c.rbuf)
+	c.rbuf = c.rbuf[n:]
+	return n, nil
+}
+
+func appendUint32(b []byte, x uint32) []byte {
+	return append(b, byte(x>>24), byte(x>>16), byte(x>>8), byte(x))
+}
+
+func readUint32(b []byte) uint32 {
+	return uint32(b[0])<<24 | uint32(b[1])<<16 | uint32(b[2])<<8 | uint32(b[3])
+}
+
+func readUint24(b []byte) int {
+	return int(b[0])<<16 | int(b[1])<<8 | int(b[2])
+}
+
+func readUint16(b []byte) uint16 {
+	return uint16(b[0])<<8 | uint16(b[1])
+}
+
+func checkWrittenHeader(gotHdr []byte, size int, ft FrameType, f Flag, streamID uint32) []error {
+	var errors []error
+	if gotSize := readUint24(gotHdr[0:3]); gotSize != size {
+		errors = append(errors, fmt.Errorf("Size: got %d, want %d", gotSize, size))
+	}
+	if gotType := FrameType(gotHdr[3]); gotType != ft {
+		errors = append(errors, fmt.Errorf("Type: got %#x, want %#x", gotType, ft))
+	}
+	if gotFlag := Flag(gotHdr[4]); gotFlag != f {
+		errors = append(errors, fmt.Errorf("Flags: got %#x, want %#x", gotFlag, f))
+	}
+	if gotSID := readUint32(gotHdr[5:9]); gotSID != streamID {
+		errors = append(errors, fmt.Errorf("StreamID: got %d, want %d", gotSID, streamID))
+	}
+	return errors
+}
+
+func checkReadHeader(gotHdr *FrameHeader, size int, ft FrameType, f Flag, streamID uint32) []error {
+	var errors []error
+	if gotHdr.Size != uint32(size) {
+		errors = append(errors, fmt.Errorf("Size: got %d, want %d", gotHdr.Size, size))
+	}
+	if gotHdr.Type != ft {
+		errors = append(errors, fmt.Errorf("Type: got %#x, want %#x", gotHdr.Type, ft))
+	}
+	if gotHdr.Flags != f {
+		errors = append(errors, fmt.Errorf("Flags: got %#x, want %#x", gotHdr.Flags, f))
+	}
+	if gotHdr.StreamID != streamID {
+		errors = append(errors, fmt.Errorf("StreamID: got %d, want %d", gotHdr.StreamID, streamID))
+	}
+	return errors
+}
+
+func (s) TestBridge_ReadFrame_Data(t *testing.T) {
+	c := &testConn{}
+	recvData := "test data"
+	c.rbuf = append(c.rbuf, 0, 0, byte(len(recvData)), byte(FrameTypeData), byte(FlagDataEndStream))
+	c.rbuf = appendUint32(c.rbuf, 1)
+	c.rbuf = append(c.rbuf, []byte(recvData)...)
```

Comment:
This initial setup is part of most tests (though not all the same). Is it possible to write a helper to set this up so that every single test need not start with a bunch of appends, where someone reading the test isn't exactly sure what is happening.

Reply:
I think you left a similar comment before. These lines are essentially writing a frame, so doing a helper function would be kind of like writing a framer (?). I was asking if this is okay for a test, because to me it looked like a framer testing against other framer. If you think it is valid, a helper function is definetely doable.

Reply:
Ok, fair enough. Just a simple one line comment saying something like `// Constructing a data frame` or something basic that gives enough information to the reader so that they don't have to focus on exactly what bit fiddling is happening in these few lines.

Reply:
Done.